Lack of stubs in Cipher.encryptor and Cipher.decryptor

the Cipher.encryptor() and Cipher.decryptor() functions provide no static type hints, this limits the amount of static typing that can be applied around the encryptor and decryptor objects

example line missing type hints:
https://github.com/pyca/cryptography/blob/686185003b03c9611fd31ba409d90d8080130f36/src/cryptography/hazmat/primitives/ciphers/base.py#L103

Hmm, you're right. I actually explicitly tested that using a variation of some of the sample code in the PyCA docs  and mypy didn't complain, but I see now that's because some of that code was missing type information. When I added that, I got the errors.

Here's the new test code:

```python
import os
from typing import Tuple

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

def encrypt(key: bytes, plaintext: bytes,
            associated_data: bytes) -> Tuple[bytes, bytes, bytes]:
    iv = os.urandom(12)

    encryptor = Cipher(algorithms.AES(key), modes.GCM(iv)).encryptor()
    encryptor.authenticate_additional_data(associated_data)
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()

    return iv, ciphertext, encryptor.tag

def decrypt(key: bytes, associated_data: bytes, iv: bytes,
            ciphertext: bytes, tag: bytes) -> bytes:
    decryptor = Cipher(algorithms.AES(key), modes.GCM(iv, tag)).decryptor()
    decryptor.authenticate_additional_data(associated_data)
    return decryptor.update(ciphertext) + decryptor.finalize()

key = os.urandom(32)
iv, ciphertext, tag = encrypt(key, b'a secret message!', b'associated data')
print(decrypt(key, b'associated data', iv, ciphertext, tag))
```

This now generates the errors:

```
t.py:11: error: "CipherContext" has no attribute "authenticate_additional_data"
t.py:14: error: "CipherContext" has no attribute "tag"
t.py:19: error: "CipherContext" has no attribute "authenticate_additional_data"
```

I tried using both typing.Union and \@typing.overload, but neither quite does the job. The Union causes it to complain because it needs both of the listed types to implement whatever methods are used in the calling code, and the \@overload doesn't work because the functions differ only in their return type and not the arguments (which contains only "self" in this case). So, it always matches the first overload.

--- a/src/cryptography/hazmat/backends/openssl/backend.py
+++ b/src/cryptography/hazmat/backends/openssl/backend.py
@@ -415,10 +415,14 @@ class Backend(BackendInterface):
                 SM4, mode_cls, GetCipherByName("sm4-{mode.name}")
             )
 
-    def create_symmetric_encryption_ctx(self, cipher, mode):
+    def create_symmetric_encryption_ctx(
+        self, cipher: CipherAlgorithm, mode: Mode
+    ) -> _CipherContext:
         return _CipherContext(self, cipher, mode, _CipherContext._ENCRYPT)
 
-    def create_symmetric_decryption_ctx(self, cipher, mode):
+    def create_symmetric_decryption_ctx(
+        self, cipher: CipherAlgorithm, mode: Mode
+    ) -> _CipherContext:
         return _CipherContext(self, cipher, mode, _CipherContext._DECRYPT)
 
     def pbkdf2_hmac_supported(self, algorithm: hashes.HashAlgorithm) -> bool:

--- a/src/cryptography/hazmat/backends/openssl/ciphers.py
+++ b/src/cryptography/hazmat/backends/openssl/ciphers.py
@@ -4,7 +4,6 @@
 
 import typing
 
-from cryptography import utils
 from cryptography.exceptions import InvalidTag, UnsupportedAlgorithm, _Reasons
 from cryptography.hazmat.primitives import ciphers
 from cryptography.hazmat.primitives.ciphers import algorithms, modes
@@ -14,10 +13,6 @@ if typing.TYPE_CHECKING:
     from cryptography.hazmat.backends.openssl.backend import Backend
 
 
-@utils.register_interface(ciphers.CipherContext)
-@utils.register_interface(ciphers.AEADCipherContext)
-@utils.register_interface(ciphers.AEADEncryptionContext)
-@utils.register_interface(ciphers.AEADDecryptionContext)
 class _CipherContext(object):
     _ENCRYPT = 1
     _DECRYPT = 0

--- a/src/cryptography/hazmat/backends/openssl/ciphers.py
+++ b/src/cryptography/hazmat/backends/openssl/ciphers.py
@@ -7,7 +7,7 @@ import typing
 from cryptography import utils
 from cryptography.exceptions import InvalidTag, UnsupportedAlgorithm, _Reasons
 from cryptography.hazmat.primitives import ciphers
-from cryptography.hazmat.primitives.ciphers import modes
+from cryptography.hazmat.primitives.ciphers import algorithms, modes
 
 
 if typing.TYPE_CHECKING:
@@ -71,7 +71,7 @@ class _CipherContext(object):
             iv_nonce = self._backend._ffi.from_buffer(mode.tweak)
         elif isinstance(mode, modes.ModeWithNonce):
             iv_nonce = self._backend._ffi.from_buffer(mode.nonce)
-        elif isinstance(cipher, modes.ModeWithNonce):
+        elif isinstance(cipher, algorithms.ChaCha20):
             iv_nonce = self._backend._ffi.from_buffer(cipher.nonce)
         else:
             iv_nonce = self._backend._ffi.NULL

--- a/src/cryptography/hazmat/primitives/ciphers/algorithms.py
+++ b/src/cryptography/hazmat/primitives/ciphers/algorithms.py
@@ -8,7 +8,6 @@ from cryptography.hazmat.primitives.ciphers import (
     BlockCipherAlgorithm,
     CipherAlgorithm,
 )
-from cryptography.hazmat.primitives.ciphers.modes import ModeWithNonce
 
 
 def _verify_key_size(algorithm: CipherAlgorithm, key: bytes) -> bytes:
@@ -133,7 +132,7 @@ class SEED(CipherAlgorithm, BlockCipherAlgorithm):
         return len(self.key) * 8
 
 
-class ChaCha20(CipherAlgorithm, ModeWithNonce):
+class ChaCha20(CipherAlgorithm):
     name = "ChaCha20"
     key_sizes = frozenset([256])
 

--- a/src/cryptography/hazmat/primitives/ciphers/base.py
+++ b/src/cryptography/hazmat/primitives/ciphers/base.py
@@ -15,6 +15,12 @@ from cryptography.hazmat.primitives._cipheralgorithm import CipherAlgorithm
 from cryptography.hazmat.primitives.ciphers import modes
 
 
+if typing.TYPE_CHECKING:
+    from cryptography.hazmat.backends.openssl.ciphers import (
+        _CipherContext as _BackendCipherContext,
+    )
+
+
 class CipherContext(metaclass=abc.ABCMeta):
     @abc.abstractmethod
     def update(self, data: bytes) -> bytes:
@@ -133,7 +139,11 @@ class Cipher(typing.Generic[Mode]):
         )
         return self._wrap_ctx(ctx, encrypt=False)
 
-    def _wrap_ctx(self, ctx, encrypt):
+    def _wrap_ctx(
+        self, ctx: "_BackendCipherContext", encrypt: bool
+    ) -> typing.Union[
+        AEADEncryptionContext, AEADDecryptionContext, CipherContext
+    ]:
         if isinstance(self.mode, modes.ModeWithAuthenticationTag):
             if encrypt:
                 return _AEADEncryptionContext(ctx)
@@ -155,7 +165,9 @@ _CIPHER_TYPE = Cipher[
 
 
 class _CipherContext(CipherContext):
-    def __init__(self, ctx):
+    _ctx: typing.Optional["_BackendCipherContext"]
+
+    def __init__(self, ctx: "_BackendCipherContext") -> None:
         self._ctx = ctx
 
     def update(self, data: bytes) -> bytes:
@@ -177,7 +189,10 @@ class _CipherContext(CipherContext):
 
 
 class _AEADCipherContext(AEADCipherContext):
-    def __init__(self, ctx):
+    _ctx: typing.Optional["_BackendCipherContext"]
+    _tag: typing.Optional[bytes]
+
+    def __init__(self, ctx: "_BackendCipherContext") -> None:
         self._ctx = ctx
         self._bytes_processed = 0
         self._aad_bytes_processed = 0
@@ -198,10 +213,14 @@ class _AEADCipherContext(AEADCipherContext):
 
     def update(self, data: bytes) -> bytes:
         self._check_limit(len(data))
+        # mypy needs this assert even though _check_limit already checked
+        assert self._ctx is not None
         return self._ctx.update(data)
 
     def update_into(self, data: bytes, buf: bytes) -> int:
         self._check_limit(len(data))
+        # mypy needs this assert even though _check_limit already checked
+        assert self._ctx is not None
         return self._ctx.update_into(data, buf)
 
     def finalize(self) -> bytes:

--- a/src/cryptography/hazmat/primitives/ciphers/base.py
+++ b/src/cryptography/hazmat/primitives/ciphers/base.py
@@ -6,7 +6,6 @@
 import abc
 import typing
 
-from cryptography import utils
 from cryptography.exceptions import (
     AlreadyFinalized,
     AlreadyUpdated,
@@ -38,7 +37,7 @@ class CipherContext(metaclass=abc.ABCMeta):
         """
 
 
-class AEADCipherContext(metaclass=abc.ABCMeta):
+class AEADCipherContext(CipherContext, metaclass=abc.ABCMeta):
     @abc.abstractmethod
     def authenticate_additional_data(self, data: bytes) -> None:
         """
@@ -46,7 +45,7 @@ class AEADCipherContext(metaclass=abc.ABCMeta):
         """
 
 
-class AEADDecryptionContext(metaclass=abc.ABCMeta):
+class AEADDecryptionContext(AEADCipherContext, metaclass=abc.ABCMeta):
     @abc.abstractmethod
     def finalize_with_tag(self, tag: bytes) -> bytes:
         """
@@ -55,7 +54,7 @@ class AEADDecryptionContext(metaclass=abc.ABCMeta):
         """
 
 
-class AEADEncryptionContext(metaclass=abc.ABCMeta):
+class AEADEncryptionContext(AEADCipherContext, metaclass=abc.ABCMeta):
     @abc.abstractproperty
     def tag(self) -> bytes:
         """
@@ -64,11 +63,16 @@ class AEADEncryptionContext(metaclass=abc.ABCMeta):
         """
 
 
-class Cipher(object):
+Mode = typing.TypeVar(
+    "Mode", bound=typing.Optional[modes.Mode], covariant=True
+)
+
+
+class Cipher(typing.Generic[Mode]):
     def __init__(
         self,
         algorithm: CipherAlgorithm,
-        mode: typing.Optional[modes.Mode],
+        mode: Mode,
         backend: typing.Any = None,
     ):
 
@@ -76,10 +80,25 @@ class Cipher(object):
             raise TypeError("Expected interface of CipherAlgorithm.")
 
         if mode is not None:
+            # mypy needs this assert to narrow the type from our generic
+            # type. Maybe it won't some time in the future.
+            assert isinstance(mode, modes.Mode)
             mode.validate_for_algorithm(algorithm)
 
         self.algorithm = algorithm
         self.mode = mode
+
+    @typing.overload
+    def encryptor(
+        self: "Cipher[modes.ModeWithAuthenticationTag]",
+    ) -> AEADEncryptionContext:
+        ...
+
+    @typing.overload
+    def encryptor(
+        self: "_CIPHER_TYPE",
+    ) -> CipherContext:
+        ...
 
     def encryptor(self):
         if isinstance(self.mode, modes.ModeWithAuthenticationTag):
@@ -94,6 +113,18 @@ class Cipher(object):
         )
         return self._wrap_ctx(ctx, encrypt=True)
 
+    @typing.overload
+    def decryptor(
+        self: "Cipher[modes.ModeWithAuthenticationTag]",
+    ) -> AEADDecryptionContext:
+        ...
+
+    @typing.overload
+    def decryptor(
+        self: "_CIPHER_TYPE",
+    ) -> CipherContext:
+        ...
+
     def decryptor(self):
         from cryptography.hazmat.backends.openssl.backend import backend
 
@@ -107,13 +138,23 @@ class Cipher(object):
             if encrypt:
                 return _AEADEncryptionContext(ctx)
             else:
-                return _AEADCipherContext(ctx)
+                return _AEADDecryptionContext(ctx)
         else:
             return _CipherContext(ctx)
 
 
-@utils.register_interface(CipherContext)
-class _CipherContext(object):
+_CIPHER_TYPE = Cipher[
+    typing.Union[
+        modes.ModeWithNonce,
+        modes.ModeWithTweak,
+        None,
+        modes.ECB,
+        modes.ModeWithInitializationVector,
+    ]
+]
+
+
+class _CipherContext(CipherContext):
     def __init__(self, ctx):
         self._ctx = ctx
 
@@ -135,10 +176,7 @@ class _CipherContext(object):
         return data
 
 
-@utils.register_interface(AEADCipherContext)
-@utils.register_interface(CipherContext)
-@utils.register_interface(AEADDecryptionContext)
-class _AEADCipherContext(object):
+class _AEADCipherContext(AEADCipherContext):
     def __init__(self, ctx):
         self._ctx = ctx
         self._bytes_processed = 0
@@ -174,14 +212,6 @@ class _AEADCipherContext(object):
         self._ctx = None
         return data
 
-    def finalize_with_tag(self, tag: bytes) -> bytes:
-        if self._ctx is None:
-            raise AlreadyFinalized("Context was already finalized.")
-        data = self._ctx.finalize_with_tag(tag)
-        self._tag = self._ctx.tag
-        self._ctx = None
-        return data
-
     def authenticate_additional_data(self, data: bytes) -> None:
         if self._ctx is None:
             raise AlreadyFinalized("Context was already finalized.")
@@ -199,8 +229,17 @@ class _AEADCipherContext(object):
         self._ctx.authenticate_additional_data(data)
 
 
-@utils.register_interface(AEADEncryptionContext)
-class _AEADEncryptionContext(_AEADCipherContext):
+class _AEADDecryptionContext(_AEADCipherContext, AEADDecryptionContext):
+    def finalize_with_tag(self, tag: bytes) -> bytes:
+        if self._ctx is None:
+            raise AlreadyFinalized("Context was already finalized.")
+        data = self._ctx.finalize_with_tag(tag)
+        self._tag = self._ctx.tag
+        self._ctx = None
+        return data
+
+
+class _AEADEncryptionContext(_AEADCipherContext, AEADEncryptionContext):
     @property
     def tag(self) -> bytes:
         if self._ctx is not None:

--- a/src/cryptography/hazmat/primitives/ciphers/modes.py
+++ b/src/cryptography/hazmat/primitives/ciphers/modes.py
@@ -29,7 +29,7 @@ class Mode(metaclass=abc.ABCMeta):
         """
 
 
-class ModeWithInitializationVector(metaclass=abc.ABCMeta):
+class ModeWithInitializationVector(Mode, metaclass=abc.ABCMeta):
     @abc.abstractproperty
     def initialization_vector(self) -> bytes:
         """
@@ -37,7 +37,7 @@ class ModeWithInitializationVector(metaclass=abc.ABCMeta):
         """
 
 
-class ModeWithTweak(metaclass=abc.ABCMeta):
+class ModeWithTweak(Mode, metaclass=abc.ABCMeta):
     @abc.abstractproperty
     def tweak(self) -> bytes:
         """
@@ -45,7 +45,7 @@ class ModeWithTweak(metaclass=abc.ABCMeta):
         """
 
 
-class ModeWithNonce(metaclass=abc.ABCMeta):
+class ModeWithNonce(Mode, metaclass=abc.ABCMeta):
     @abc.abstractproperty
     def nonce(self) -> bytes:
         """
@@ -53,7 +53,7 @@ class ModeWithNonce(metaclass=abc.ABCMeta):
         """
 
 
-class ModeWithAuthenticationTag(metaclass=abc.ABCMeta):
+class ModeWithAuthenticationTag(Mode, metaclass=abc.ABCMeta):
     @abc.abstractproperty
     def tag(self) -> typing.Optional[bytes]:
         """
@@ -89,7 +89,7 @@ def _check_iv_and_key_length(self, algorithm):
     _check_iv_length(self, algorithm)
 
 
-class CBC(Mode, ModeWithInitializationVector):
+class CBC(ModeWithInitializationVector):
     name = "CBC"
 
     def __init__(self, initialization_vector: bytes):
@@ -103,7 +103,7 @@ class CBC(Mode, ModeWithInitializationVector):
     validate_for_algorithm = _check_iv_and_key_length
 
 
-class XTS(Mode, ModeWithTweak):
+class XTS(ModeWithTweak):
     name = "XTS"
 
     def __init__(self, tweak: bytes):
@@ -132,7 +132,7 @@ class ECB(Mode):
     validate_for_algorithm = _check_aes_key_length
 
 
-class OFB(Mode, ModeWithInitializationVector):
+class OFB(ModeWithInitializationVector):
     name = "OFB"
 
     def __init__(self, initialization_vector: bytes):
@@ -146,7 +146,7 @@ class OFB(Mode, ModeWithInitializationVector):
     validate_for_algorithm = _check_iv_and_key_length
 
 
-class CFB(Mode, ModeWithInitializationVector):
+class CFB(ModeWithInitializationVector):
     name = "CFB"
 
     def __init__(self, initialization_vector: bytes):
@@ -160,7 +160,7 @@ class CFB(Mode, ModeWithInitializationVector):
     validate_for_algorithm = _check_iv_and_key_length
 
 
-class CFB8(Mode, ModeWithInitializationVector):
+class CFB8(ModeWithInitializationVector):
     name = "CFB8"
 
     def __init__(self, initialization_vector: bytes):
@@ -174,7 +174,7 @@ class CFB8(Mode, ModeWithInitializationVector):
     validate_for_algorithm = _check_iv_and_key_length
 
 
-class CTR(Mode, ModeWithNonce):
+class CTR(ModeWithNonce):
     name = "CTR"
 
     def __init__(self, nonce: bytes):
@@ -190,7 +190,7 @@ class CTR(Mode, ModeWithNonce):
         _check_nonce_length(self.nonce, self.name, algorithm)
 
 
-class GCM(Mode, ModeWithInitializationVector, ModeWithAuthenticationTag):
+class GCM(ModeWithInitializationVector, ModeWithAuthenticationTag):
     name = "GCM"
     _MAX_ENCRYPTED_BYTES = (2 ** 39 - 256) // 8
     _MAX_AAD_BYTES = (2 ** 64) // 8

--- a/src/cryptography/hazmat/primitives/keywrap.py
+++ b/src/cryptography/hazmat/primitives/keywrap.py
@@ -120,10 +120,10 @@ def aes_key_unwrap_with_padding(
     if len(wrapped_key) == 16:
         # RFC 5649 - 4.2 - exactly two 64-bit blocks
         decryptor = Cipher(AES(wrapping_key), ECB()).decryptor()
-        b = decryptor.update(wrapped_key)
+        out = decryptor.update(wrapped_key)
         assert decryptor.finalize() == b""
-        a = b[:8]
-        data = b[8:]
+        a = out[:8]
+        data = out[8:]
         n = 1
     else:
         r = [wrapped_key[i : i + 8] for i in range(0, len(wrapped_key), 8)]

--- a/tests/hazmat/primitives/test_aes_gcm.py
+++ b/tests/hazmat/primitives/test_aes_gcm.py
@@ -71,9 +71,11 @@ class TestAESModeGCM(object):
             modes.GCM(b"\x01" * 16),
             backend=backend,
         ).encryptor()
-        encryptor._bytes_processed = modes.GCM._MAX_ENCRYPTED_BYTES - 16
+        new_max = modes.GCM._MAX_ENCRYPTED_BYTES - 16
+        encryptor._bytes_processed = new_max  # type: ignore[attr-defined]
         encryptor.update(b"0" * 16)
-        assert encryptor._bytes_processed == modes.GCM._MAX_ENCRYPTED_BYTES
+        max = modes.GCM._MAX_ENCRYPTED_BYTES
+        assert encryptor._bytes_processed == max  # type: ignore[attr-defined]
         with pytest.raises(ValueError):
             encryptor.update(b"0")
 
@@ -83,9 +85,13 @@ class TestAESModeGCM(object):
             modes.GCM(b"\x01" * 16),
             backend=backend,
         ).encryptor()
-        encryptor._aad_bytes_processed = modes.GCM._MAX_AAD_BYTES - 16
+        new_max = modes.GCM._MAX_AAD_BYTES - 16
+        encryptor._aad_bytes_processed = new_max  # type: ignore[attr-defined]
         encryptor.authenticate_additional_data(b"0" * 16)
-        assert encryptor._aad_bytes_processed == modes.GCM._MAX_AAD_BYTES
+        max = modes.GCM._MAX_AAD_BYTES
+        assert (
+            encryptor._aad_bytes_processed == max  # type: ignore[attr-defined]
+        )
         with pytest.raises(ValueError):
             encryptor.authenticate_additional_data(b"0")
 
@@ -96,11 +102,11 @@ class TestAESModeGCM(object):
             backend=backend,
         ).encryptor()
         encryptor.update(b"0" * 8)
-        assert encryptor._bytes_processed == 8
+        assert encryptor._bytes_processed == 8  # type: ignore[attr-defined]
         encryptor.update(b"0" * 7)
-        assert encryptor._bytes_processed == 15
+        assert encryptor._bytes_processed == 15  # type: ignore[attr-defined]
         encryptor.update(b"0" * 18)
-        assert encryptor._bytes_processed == 33
+        assert encryptor._bytes_processed == 33  # type: ignore[attr-defined]
 
     def test_gcm_aad_increments(self, backend):
         encryptor = base.Cipher(
@@ -109,9 +115,13 @@ class TestAESModeGCM(object):
             backend=backend,
         ).encryptor()
         encryptor.authenticate_additional_data(b"0" * 8)
-        assert encryptor._aad_bytes_processed == 8
+        assert (
+            encryptor._aad_bytes_processed == 8  # type: ignore[attr-defined]
+        )
         encryptor.authenticate_additional_data(b"0" * 18)
-        assert encryptor._aad_bytes_processed == 26
+        assert (
+            encryptor._aad_bytes_processed == 26  # type: ignore[attr-defined]
+        )
 
     def test_gcm_tag_decrypt_none(self, backend):
         key = binascii.unhexlify(b"5211242698bed4774a090620a6ca56f3")

--- a/tests/hazmat/primitives/test_ciphers.py
+++ b/tests/hazmat/primitives/test_ciphers.py
@@ -325,14 +325,18 @@ class TestCipherUpdateInto(object):
         c = ciphers.Cipher(AES(key), modes.ECB(), backend)
         encryptor = c.encryptor()
         # Lower max chunk size so we can test chunking
-        monkeypatch.setattr(encryptor._ctx, "_MAX_CHUNK_SIZE", 40)
+        monkeypatch.setattr(
+            encryptor._ctx, "_MAX_CHUNK_SIZE", 40  # type: ignore[attr-defined]
+        )
         buf = bytearray(527)
         pt = b"abcdefghijklmnopqrstuvwxyz012345" * 16  # 512 bytes
         processed = encryptor.update_into(pt, buf)
         assert processed == 512
         decryptor = c.decryptor()
         # Change max chunk size to verify alternate boundaries don't matter
-        monkeypatch.setattr(decryptor._ctx, "_MAX_CHUNK_SIZE", 73)
+        monkeypatch.setattr(
+            decryptor._ctx, "_MAX_CHUNK_SIZE", 73  # type: ignore[attr-defined]
+        )
         decbuf = bytearray(527)
         decprocessed = decryptor.update_into(buf[:processed], decbuf)
         assert decbuf[:decprocessed] == pt
@@ -344,4 +348,7 @@ class TestCipherUpdateInto(object):
         key = b"\x00" * 16
         c = ciphers.Cipher(AES(key), modes.ECB(), backend)
         encryptor = c.encryptor()
-        backend._ffi.new("int *", encryptor._ctx._MAX_CHUNK_SIZE)
+        backend._ffi.new(
+            "int *",
+            encryptor._ctx._MAX_CHUNK_SIZE,  # type: ignore[attr-defined]
+        )

--- a/tests/hazmat/primitives/utils.py
+++ b/tests/hazmat/primitives/utils.py
@@ -106,6 +106,12 @@ def aead_test(backend, cipher_factory, mode_factory, params):
         # hex encoded.
         pytest.skip("Non-96-bit IVs unsupported in FIPS mode.")
 
+    mode = mode_factory(
+        binascii.unhexlify(params["iv"]),
+        binascii.unhexlify(params["tag"]),
+        len(binascii.unhexlify(params["tag"])),
+    )
+    assert isinstance(mode, GCM)
     if params.get("pt") is not None:
         plaintext = binascii.unhexlify(params["pt"])
     ciphertext = binascii.unhexlify(params["ct"])
@@ -113,11 +119,7 @@ def aead_test(backend, cipher_factory, mode_factory, params):
     if params.get("fail") is True:
         cipher = Cipher(
             cipher_factory(binascii.unhexlify(params["key"])),
-            mode_factory(
-                binascii.unhexlify(params["iv"]),
-                binascii.unhexlify(params["tag"]),
-                len(binascii.unhexlify(params["tag"])),
-            ),
+            mode,
             backend,
         )
         decryptor = cipher.decryptor()
@@ -291,9 +293,11 @@ def generate_aead_exception_test(cipher_factory, mode_factory):
 
 
 def aead_exception_test(backend, cipher_factory, mode_factory):
+    mode = mode_factory(binascii.unhexlify(b"0" * 24))
+    assert isinstance(mode, GCM)
     cipher = Cipher(
         cipher_factory(binascii.unhexlify(b"0" * 32)),
-        mode_factory(binascii.unhexlify(b"0" * 24)),
+        mode,
         backend,
     )
     encryptor = cipher.encryptor()
@@ -309,15 +313,18 @@ def aead_exception_test(backend, cipher_factory, mode_factory):
         encryptor.update(b"b" * 16)
     with pytest.raises(AlreadyFinalized):
         encryptor.finalize()
+
+    mode2 = mode_factory(binascii.unhexlify(b"0" * 24), b"0" * 16)
+    assert isinstance(mode2, GCM)
     cipher = Cipher(
         cipher_factory(binascii.unhexlify(b"0" * 32)),
-        mode_factory(binascii.unhexlify(b"0" * 24), b"0" * 16),
+        mode2,
         backend,
     )
     decryptor = cipher.decryptor()
     decryptor.update(b"a" * 16)
     with pytest.raises(AttributeError):
-        decryptor.tag
+        decryptor.tag  # type: ignore[attr-defined]
 
 
 def generate_aead_tag_exception_test(cipher_factory, mode_factory):

